File: optimization-sorting-algorithms-problems-pro/utils/arrayGenerator.js
```javascript
/**
 * utils/arrayGenerator.js
 *
 * This file provides utility functions for generating various types of arrays,
 * useful for testing and benchmarking sorting algorithms.
 */

/**
 * Generates an array of random integers.
 * @param {number} size The number of elements in the array.
 * @param {number} [min=0] The minimum value for elements.
 * @param {number} [max=1000] The maximum value for elements.
 * @returns {Array<number>} An array of random integers.
 */
function generateRandomArray(size, min = 0, max = 1000) {
    if (size <= 0) return [];
    const arr = new Array(size);
    for (let i = 0; i < size; i++) {
        arr[i] = Math.floor(Math.random() * (max - min + 1)) + min;
    }
    return arr;
}

/**
 * Generates an array that is already sorted in ascending order.
 * @param {number} size The number of elements in the array.
 * @param {number} [start=0] The starting value.
 * @param {number} [step=1] The increment between elements.
 * @returns {Array<number>} A sorted array.
 */
function generateSortedArray(size, start = 0, step = 1) {
    if (size <= 0) return [];
    const arr = new Array(size);
    for (let i = 0; i < size; i++) {
        arr[i] = start + i * step;
    }
    return arr;
}

/**
 * Generates an array sorted in descending (reverse) order.
 * @param {number} size The number of elements in the array.
 * @param {number} [start=0] The starting value (will be the largest).
 * @param {number} [step=1] The decrement between elements.
 * @returns {Array<number>} A reverse sorted array.
 */
function generateReverseSortedArray(size, start = 0, step = 1) {
    if (size <= 0) return [];
    const arr = new Array(size);
    for (let i = 0; i < size; i++) {
        arr[i] = start - i * step;
    }
    return arr;
}

/**
 * Generates an array with many duplicate values.
 * @param {number} size The number of elements in the array.
 * @param {number} [numUnique=10] The number of unique values to choose from.
 * @param {number} [min=0] The minimum value for unique elements.
 * @returns {Array<number>} An array with many duplicates.
 */
function generateDuplicatesArray(size, numUnique = 10, min = 0) {
    if (size <= 0) return [];
    const arr = new Array(size);
    const uniqueValues = generateRandomArray(numUnique, min, min + numUnique * 2);
    for (let i = 0; i < size; i++) {
        arr[i] = uniqueValues[Math.floor(Math.random() * numUnique)];
    }
    return arr;
}

/**
 * Generates an array that is mostly sorted, but with a few elements out of place.
 * @param {number} size The number of elements in the array.
 * @param {number} [swaps=2] The number of random swaps to perform on a sorted array.
 * @returns {Array<number>} A nearly sorted array.
 */
function generateNearlySortedArray(size, swaps = 2) {
    if (size <= 0) return [];
    const arr = generateSortedArray(size);
    for (let i = 0; i < swaps; i++) {
        const idx1 = Math.floor(Math.random() * size);
        const idx2 = Math.floor(Math.random() * size);
        if (idx1 !== idx2) {
            const temp = arr[idx1];
            arr[idx1] = arr[idx2];
            arr[idx2] = temp;
        }
    }
    return arr;
}

/**
 * Generates an array with all elements being the same value.
 * @param {number} size The number of elements.
 * @param {number} [value=5] The value for all elements.
 * @returns {Array<number>} An array filled with a single value.
 */
function generateAllSameArray(size, value = 5) {
    if (size <= 0) return [];
    return new Array(size).fill(value);
}

module.exports = {
    generateRandomArray,
    generateSortedArray,
    generateReverseSortedArray,
    generateDuplicatesArray,
    generateNearlySortedArray,
    generateAllSameArray
};
```

File: data-structures-sorting-algorithms-enhanced/src/main/java/com/example/sorting/utils/ArrayGenerator.java
```java
package com.example.sorting.utils;

import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Utility class for generating various types of integer arrays for testing and benchmarking sorting algorithms.
 */
public class ArrayGenerator {

    private static final Random RANDOM = ThreadLocalRandom.current();

    /**
     * Generates an array of a specified size with random integer values.
     *
     * @param size The size of the array to generate.
     * @param min  The minimum possible value for elements.
     * @param max  The maximum possible value for elements.
     * @return A new integer array with random values.
     * @throws IllegalArgumentException if size is negative or max is less than min.
     */
    public static int[] generateRandomArray(int size, int min, int max) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        if (min > max) {
            throw new IllegalArgumentException("Min value cannot be greater than max value.");
        }

        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = RANDOM.nextInt(max - min + 1) + min;
        }
        return arr;
    }

    /**
     * Generates a sorted array of a specified size with distinct values.
     *
     * @param size The size of the array.
     * @return A new sorted integer array.
     * @throws IllegalArgumentException if size is negative.
     */
    public static int[] generateSortedArray(int size) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = i; // Simple ascending sequence
        }
        return arr;
    }

    /**
     * Generates a reverse sorted array of a specified size with distinct values.
     *
     * @param size The size of the array.
     * @return A new reverse sorted integer array.
     * @throws IllegalArgumentException if size is negative.
     */
    public static int[] generateReverseSortedArray(int size) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = size - 1 - i; // Simple descending sequence
        }
        return arr;
    }

    /**
     * Generates an array with many duplicate values.
     *
     * @param size The size of the array.
     * @param numDistinct The number of distinct values to use.
     * @return An array with many duplicates.
     * @throws IllegalArgumentException if size is negative or numDistinct is non-positive.
     */
    public static int[] generateArrayWithDuplicates(int size, int numDistinct) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        if (numDistinct <= 0) {
            throw new IllegalArgumentException("Number of distinct elements must be positive.");
        }
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = RANDOM.nextInt(numDistinct); // Values will be between 0 and numDistinct-1
        }
        return arr;
    }

    /**
     * Generates an array of a specified size with elements all having the same value.
     *
     * @param size The size of the array.
     * @param value The value to fill the array with.
     * @return A new array with all elements equal to 'value'.
     * @throws IllegalArgumentException if size is negative.
     */
    public static int[] generateConstantArray(int size, int value) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = value;
        }
        return arr;
    }
}
```